micro-rdk/src/common/data_manager.rs

```diff
@@ -222,6 +202,106 @@ where
             .map(|coll| Ok((coll.resource_method_key(), coll.call_method()?)))
             .collect()
     }
+
+    pub fn get_sync_task(&self) -> DataSyncTask<StoreType> {
+        let resource_method_keys: Vec<ResourceMethodKey> = self
+            .collectors
+            .iter()
+            .map(|coll| coll.resource_method_key())
+            .collect();
+        DataSyncTask {
+            store: self.store.clone(),
+            resource_method_keys,
+            sync_interval: self.sync_interval,
+            part_id: self.part_id(),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum DataSyncError {
+    #[error(transparent)]
+    DataStoreError(#[from] DataStoreError),
+    #[error(transparent)]
+    MessageDecodingError(#[from] prost::DecodeError),
+}
+
+pub struct DataSyncTask<StoreType> {
+    store: Rc<AsyncMutex<StoreType>>,
+    resource_method_keys: Vec<ResourceMethodKey>,
+    sync_interval: Duration,
+    part_id: String,
+}
+
+impl<StoreType> DataSyncTask<StoreType>
+where
+    StoreType: DataStore,
+{
+    async fn read_messages_for_collector(
+        &self,
+        collector_key: &ResourceMethodKey,
+    ) -> Result<Vec<SensorData>, DataSyncError> {
+        let store_lock = self.store.lock().await;
+        let mut raw_messages: Vec<BytesMut> = vec![];
+        let mut reader = store_lock.get_reader(collector_key)?;
+        loop {
+            let next_message = reader.read_next_message()?;
+            if next_message.is_empty() {
+                break;
+            }
+            raw_messages.push(next_message);
+        }
+        let data: Result<Vec<SensorData>, prost::DecodeError> =
+            raw_messages.into_iter().map(SensorData::decode).collect();
+        Ok(data?)
+    }
+
+    async fn run<'b>(&mut self, app_client: &'b AppClient) -> Result<(), AppClientError> {
+        for collector_key in self.resource_method_keys.iter() {
+            let data = match self.read_messages_for_collector(collector_key).await {
```

Comment:
I think at some point we had discussed chunking this into multiple upload requests not exceeding some size. The concern being that if you come up after a long time offline and want to sync, you might have a very large amount of data to send: too big to pull into heap for a single message.

If we agree that we want that, I'd be OK with a ticket and a TODO rather than continuing to iterate this review.

Reply:
I can implement the chunking here

Reply:
actually, I'll do it in a different PR just to keep the reviewable chunks of code a little smaller, but I intend to open that PR ASAP following this one

Reply:
Sounds good to push chunking to subsequent review; resolving.